subsys/net/openthread/rpc/client/ot_rpc_dns_client.c

```diff
@@ -0,0 +1,551 @@
+			    OT_RPC_CMD_DNS_CLIENT_GET_DEFAULT_CONFIG, NRF_RPC_PACKET_TYPE_RSP);
+	}
+
+	return &default_config;
+}
+
+void otDnsClientSetDefaultConfig(otInstance *aInstance, const otDnsQueryConfig *aConfig)
+{
+	struct nrf_rpc_cbor_ctx ctx;
+	bool encoded;
+
+	ARG_UNUSED(aInstance);
+
+	NRF_RPC_CBOR_ALLOC(&ot_group, ctx, OT_RPC_DNS_QUERY_CONFIG_SIZE);
+
+	encoded = ot_rpc_encode_dns_query_config(&ctx, aConfig);
+
+	if (encoded) {
+		nrf_rpc_cbor_cmd_no_err(&ot_group, OT_RPC_CMD_DNS_CLIENT_SET_DEFAULT_CONFIG, &ctx,
+					ot_rpc_decode_void, NULL);
+	}
+}
+
+static otError resolve_or_browse(enum ot_rpc_cmd_server cmd, const char *name, const char *service,
+				 void *cb, void *data, const otDnsQueryConfig *config)
+{
+	struct nrf_rpc_cbor_ctx ctx;
+	size_t to_alloc;
+	size_t name_len;
+	size_t service_len;
+
+	if (!name) {
+		return OT_ERROR_INVALID_ARGS;
+	}
+
+	name_len = strlen(name);
+
+	to_alloc = name_len + 2;
+	to_alloc += sizeof(uintptr_t) + 1;
+	to_alloc += sizeof(uint32_t) + 1;
+
+	if (service) {
+		service_len = strlen(service);
+		to_alloc += service_len + 2;
+	}
+
+	if (config) {
+		to_alloc += OT_RPC_DNS_QUERY_CONFIG_SIZE;
+	}
+
+	NRF_RPC_CBOR_ALLOC(&ot_group, ctx, to_alloc);
+
+	nrf_rpc_encode_str(&ctx, name, name_len + 1);
+
+	if (service) {
+		nrf_rpc_encode_str(&ctx, service, service_len + 1);
+	}
+
+	nrf_rpc_encode_callback(&ctx, cb);
+	nrf_rpc_encode_uint(&ctx, (uintptr_t)data);
+
+	if (config) {
+		ot_rpc_encode_dns_query_config(&ctx, config);
+	}
+
+	nrf_rpc_cbor_cmd_rsp_no_err(&ot_group, cmd, &ctx);
+
+	return decode_ot_error(&ctx);
+}
+
+static otError fetch_strings_from_dns_response(enum ot_rpc_cmd_server cmd, const void *response,
+					       uint16_t *index, char *primary, uint16_t primary_len,
+					       char *secondary, uint16_t secondary_len)
+{
+	struct nrf_rpc_cbor_ctx ctx;
+	uint32_t error;
+	struct zcbor_string result;
+
+	if (!response || !primary || !primary_len) {
+		return OT_ERROR_INVALID_ARGS;
+	}
+
+	NRF_RPC_CBOR_ALLOC(&ot_group, ctx, 4 * (sizeof(uint32_t) + 1));
+
+	nrf_rpc_encode_uint(&ctx, (uintptr_t)(response));
+
+	if (index) {
+		nrf_rpc_encode_uint(&ctx, *index);
+	}
+
+	nrf_rpc_encode_uint(&ctx, primary_len);
+
+	if (secondary) {
+		nrf_rpc_encode_uint(&ctx, secondary_len);
+	}
+
+	nrf_rpc_cbor_cmd_rsp_no_err(&ot_group, cmd, &ctx);
+
+	if (!zcbor_uint_decode(ctx.zs, &error, sizeof(uint32_t))) {
+		goto fail;
+	}
+
+	/* Verify that we can proceed with decoding the name */
+	if (error != OT_ERROR_NONE) {
+		nrf_rpc_cbor_decoding_done(&ot_group, &ctx);
+		return error;
+	}
+
+	if (!zcbor_tstr_decode(ctx.zs, &result)) {
+		goto fail;
+	}
+
+	memcpy(primary, result.value, result.len);
+
+	if (secondary) {
+		if (!zcbor_tstr_decode(ctx.zs, &result)) {
+			goto fail;
+		}
+		memcpy(secondary, result.value, result.len);
+	}
+
+	nrf_rpc_cbor_decoding_done(&ot_group, &ctx);
+
+	return OT_ERROR_NONE;
+
+fail:
+	nrf_rpc_cbor_decoding_done(&ot_group, &ctx);
+	ot_rpc_report_decoding_error(cmd);
```

Comment:
Technically, this function is for reporting failures when decoding commands, not responses. Perhaps, we need another helper for responses.

Reply:
I will look into that